querybook/webapp/lib/python/python-provider.tsx

```diff
@@ -0,0 +1,170 @@
+}
+
+function PythonProvider({ children }: PythonProviderProps) {
+    const [status, setStatus] = useState(PythonKernelStatus.UNINITIALIZED);
+    const workerRef = useRef<Worker>(null);
+    const kernelRef = useRef<Remote<PythonKernel>>(null);
+
+    /**
+     * Terminate the kernel and clean up resources
+     */
+    const terminateKernel = useCallback(() => {
+        // Terminate worker
+        if (workerRef.current) {
+            workerRef.current.terminate();
+            workerRef.current = null;
+        }
+
+        // Release kernel proxy
+        if (kernelRef.current) {
+            kernelRef.current[releaseProxy]();
+            kernelRef.current = null;
+        }
+
+        // Reset state
+        setStatus(PythonKernelStatus.UNINITIALIZED);
+    }, [setStatus]);
+
+    const initKernel = useCallback(async () => {
+        // Don't initialize if already initializing or initialized
+        if (kernelRef.current) {
+            return;
+        }
+
+        try {
+            setStatus(PythonKernelStatus.INITIALIZING);
+
+            // Create a new worker
+            const worker = new Worker(
+                new URL('./python-worker', import.meta.url),
+                {
+                    name: 'python-worker',
+                    type: 'module',
+                }
+            );
+            workerRef.current = worker;
+
+            // Wrap the worker with comlink
+            const kernel: Remote<PythonKernel> = wrap(worker);
+            await kernel.initialize();
+            kernelRef.current = kernel;
+
+            setStatus(PythonKernelStatus.IDLE);
+        } catch (error) {
+            console.error('Error initializing Python kernel:', error);
+            setStatus(PythonKernelStatus.FAILED);
+            terminateKernel();
+        }
+    }, [setStatus, terminateKernel]);
+
+    useEffect(() => {
+        initKernel();
+
+        // cleanup on unmount
+        return () => {
+            terminateKernel();
+        };
+    }, []);
+
+    /**
+     * Get the execution count for a namespace from the kernel
+     */
+    const getExecutionCount = useCallback(async (namespaceId: number) => {
+        if (!kernelRef.current) {
+            return 0;
+        }
+        return kernelRef.current.getExecutionCount(namespaceId);
+    }, []);
+
+    /**
+     * Run Python code in the specified namespace
+     */
+    const runPython = useCallback(
+        async (
+            code: string,
+            namespaceId?: number,
+            stdoutCallback?: (text: string) => void,
+            stderrCallback?: (text: string) => void
+        ) => {
+            if (!kernelRef.current) {
+                await initKernel();
+            }
+
+            if (!kernelRef.current) {
+                stderrCallback('Failed to initialize Python kernel');
+            }
+
+            if (status !== PythonKernelStatus.IDLE) {
+                stderrCallback('Kernel is not ready to run code');
+            }
```

Comment:
are we not allowing users running multiple python cells?

Reply:
yes, we allow.

Reply:
i meant what happens if you execute cell 2 while cell 1 is running? it doesn't seem like there is a wait queue at the moment?